server/src/main/java/org/elasticsearch/snapshots/RestoreService.java

```diff
@@ -264,16 +271,42 @@ public void restoreSnapshot(final RestoreSnapshotRequest request,
                     metadataBuilder = Metadata.builder();
                 }
 
-                final List<IndexId> indexIdsInSnapshot = repositoryData.resolveIndices(indicesInSnapshot);
+                // set up feature state indices
+                Set<String> requestedFeatureStateIndexes = new HashSet<>();
+                Set<String> nonRequestedFeatureStateIndexes = new HashSet<>();
+
+                if (request.includeGlobalState() == false) {
+                    for (SnapshotFeatureInfo snapshotFeatureInfo : snapshotInfo.featureStates()) {
+                        if (Set.of(request.featureStates()).contains(snapshotFeatureInfo.getPluginName())) {
+                            requestedFeatureStateIndexes.addAll(snapshotFeatureInfo.getIndices());
+                        } else {
+                            nonRequestedFeatureStateIndexes.addAll(snapshotFeatureInfo.getIndices());
+                        }
+                    }
+                }
+
+                List<String> requestedIndicesIncludingSystem = Stream.concat(
+                    indicesInSnapshot.stream(), requestedFeatureStateIndexes.stream())
+                    .distinct()
+                    .filter(index -> nonRequestedFeatureStateIndexes.contains(index) == false)
+                    .collect(Collectors.toList());
+
+                List<Index> systemIndicesToDelete = new ArrayList<>();
```

Comment:
Since we convert this to a `HashSet` later anyway, is there any reason we don't just make it a `HashSet` to begin with?